torch/csrc/jit/codegen/cuda/lower_alias_memory.cpp

```diff
@@ -1201,28 +1263,101 @@ class AllocateReuseModifier {
+  AllocationInfoMap& allocation_info_map_;
 
   // Internal data keeping track of currently visible allocations as
   //  the pass iterate through the expr list, grouped by the stack
   //  layer of alloc ops.
-  std::vector<std::unique_ptr<AllocationInfoList>>
+  std::vector<std::unique_ptr<std::vector<AllocationInfo*>>>
       current_visible_buffer_stack_;
 
   // Marks state of current pass
   bool inner_aliasing_pass_ = true;
 };
 
+// Replace Allocate exprs as determined by the alias analysis
+class AllocationReuseModifier : private kir::ExprMutator {
+ public:
+  static std::vector<Expr*> modify(
+      const std::vector<Expr*>& exprs,
+      const AllocationInfoMap& allocation_info_map) {
+    AllocationReuseModifier modifier(exprs, allocation_info_map);
+    return modifier.exprs_;
+  }
+
+ private:
+  AllocationReuseModifier(
+      const std::vector<Expr*>& exprs,
+      const AllocationInfoMap& allocation_info_map)
+      : allocation_info_map_(allocation_info_map) {
+    traverseAndInsert(exprs);
+  }
+
+  using kir::ExprMutator::handle;
+
+  void handle(kir::Allocate* allocate) final {
+    auto maybe_alloc_info =
+        allocation_info_map_.getMaybeAllocationInfo(allocate);
+    if (!maybe_alloc_info.has_value()) {
+      return;
+    }
+
+    AllocationInfo* alloc_info_from = maybe_alloc_info.value();
+
+    auto alias_it = allocation_info_map_.getAliasMap().find(alloc_info_from);
+    if (alias_it == allocation_info_map_.getAliasMap().end()) {
+      return;
+    }
+
+    // The alias-to Allocate may be replaced with a new
+    // Allocate. Use the new one in that case.
+    kir::Allocate* alloc_expr_to =
+        getMaybeNewAllocate(alias_it->second->alloc_expr);
+
+    kir::Allocate* old_alloc = alloc_info_from->alloc_expr;
+    kir::Allocate* new_alloc = IrBuilder::create<kir::Allocate>(
+        old_alloc->buffer(),
+        old_alloc->memoryType(),
+        old_alloc->shape(),
+        old_alloc->zeroInit(),
+        alloc_expr_to);
+
+    registerReplace(old_alloc, new_alloc);
+
+    TORCH_INTERNAL_ASSERT(old2new_.emplace(old_alloc, new_alloc).second);
+
+    // TODO: Consider more robust way to keep the information map up-to-date
+    GpuLower::current()->propagateExprInfo(old_alloc, new_alloc);
+  }
+
+  kir::Allocate* getMaybeNewAllocate(kir::Allocate* allocate) const {
+    auto it = old2new_.find(allocate);
+    if (it == old2new_.end()) {
+      return allocate;
+    } else {
+      return it->second;
+    }
+  }
+
+ private:
+  const AllocationInfoMap& allocation_info_map_;
+
+  //! Keep track of new Allocate exprs
+  std::unordered_map<kir::Allocate*, kir::Allocate*> old2new_;
+};
+
 } // namespace
 
 std::vector<Expr*> reuseMemoryAllocations(const std::vector<Expr*>& exprs) {
   FUSER_PERF_SCOPE("reuseMemoryAllocations");
+
   bool debug_print = isDebugDumpEnabled(DebugDumpOption::BufferReuseInfo);
-  if (debug_print) {
-    AllocateReuseModifier::debugPrint(exprs);
-  }
-  AllocateReuseModifier::modify(exprs);
```

Comment:
Not sure why, but previously when the debug print is enabled, the analysis is executed twice. That's actually not supported as the analysis assumes anything already aliased is an output that is aliased to an input (https://github.com/csarofeen/pytorch/pull/2170/files#diff-37be3f6d58243db8255519e2eb487975a21cfe8c030c937ae169e4367f5ef6e1R682)